crates/lib/src/capture/pcap.rs

```diff
@@ -0,0 +1,386 @@
+        match info.ip_addr() {
+            (IpAddr::V4(_), IpAddr::V4(_)) => {
+                let (source, destination) = info.ipvt_by_direction();
+
+                let header_size = HEADER_SIZE_IP4 + (32 / 8);
+
+                let mut ip = MutableIpv4Packet::new(buf).unwrap();
+                ip.set_version(4);
+                ip.set_total_length((payload.len() + header_size) as u16);
+                ip.set_next_level_protocol(protocol);
+                // https://en.wikipedia.org/wiki/Internet_Protocol_version_4#Total_Length
+
+                ip.set_header_length((header_size / 4) as u8);
+                ip.set_source(source);
+                ip.set_destination(destination);
+                ip.set_payload(payload);
+                ip.set_ttl(64);
+                ip.set_flags(pnet_packet::ipv4::Ipv4Flags::DontFragment);
+
+                let mut options_writer =
+                    pnet_packet::ipv4::MutableIpv4OptionPacket::new(ip.get_options_raw_mut()).unwrap();
+                options_writer.set_copied(1);
+                options_writer.set_class(0);
+                options_writer.set_number(pnet_packet::ipv4::Ipv4OptionNumbers::SID);
+                options_writer.set_length(&[4]);
+                options_writer.set_data(&(self.state.stream_count as u16).to_be_bytes());
+
+                ip.set_checksum(pnet_packet::ipv4::checksum(&ip.to_immutable()));
+
+                (ip.packet_size(), pnet_packet::ethernet::EtherTypes::Ipv4)
+            }
+            (IpAddr::V6(_), IpAddr::V6(_)) => {
+                let (source, destination) = info.ipvt_by_direction();
+
+                let mut ip = MutableIpv6Packet::new(buf).unwrap();
+                ip.set_version(6);
+                ip.set_payload_length(payload.len() as u16);
+                ip.set_next_header(protocol);
+                ip.set_source(source);
+                ip.set_destination(destination);
+                ip.set_hop_limit(64);
+                ip.set_payload(payload);
+                ip.set_flow_label(self.state.stream_count);
+
+                (ip.packet_size(), pnet_packet::ethernet::EtherTypes::Ipv6)
+            }
+            _ => unreachable!(),
+        }
+    }
+
+    /// Encode a physical layer (ethernet) packet with a payload.
+    fn encode_ethernet_packet(
+        &self,
+        buf: &mut [u8],
+        ethertype: pnet_packet::ethernet::EtherType,
+        payload: &[u8],
+    ) -> usize {
+        let mut ethernet = MutableEthernetPacket::new(buf).unwrap();
+        ethernet.set_ethertype(ethertype);
+        ethernet.set_payload(payload);
+
+        ethernet.packet_size()
+    }
+
+    /// Write a TCP handshake.
+    pub(crate) fn write_tcp_handshake(&mut self, info: &CapturePacket) {
+        let (source_port, dest_port) = (info.local_address.port(), info.remote_address.port());
+
+        let mut info = info.clone();
+        info.direction = Direction::Send;
+        let mut buf = vec![0; PACKET_SIZE];
```

Comment:
this is true, idk if it would benefit that much as this is excluded from default although it would have benefits of avoiding dynamic allocation overhead with the vec. I would think it would look something like this

```rs
let mut buffer_array: [u8; PACKET_SIZE] = [0; PACKET_SIZE];
let buf: &mut [u8] = &mut buffer_array[..];
```

Reply:
i have done this where possible/practical in pcap